### ISSUE TITLE
Cross System Comparison

### ISSUE THREAD

**Issue by mbtools** [abapGit/abapGit#4784]
One common challenge for SAP customers is finding differences between systems. I occurred to me that abapGit could be great at comparing repos across systems. 

I see this similar to how offline repos work. However, instead of uploading the ZIP, we retrieve the ZIP via an RFC connection from another SAP system. 

AG can already serialize to ZIP so adding an RFC function to AG to get this ZIP for a given SAP package would get us really close to making this feature happen.

**Comment by larshp** [comment]
With zip, I guess this is an ad-hoc task? And this feature via zip is borderline git client.

With git, connect the two systems to a git server(abapGitServer?), push from one system, compare in the other system.

Alternatively, download the 2 zips, unzip to local pc, and use a tool like [WinMerge](https://winmerge.org/) to compare.

Plus adding the "compare zips via RFC feature", will require new screens(eg. select RFC destination) and objects, as abapGit works only client <-> server, not client <-> client

**Comment by flaiker** [comment]
So basically integrate this process into the abapGit UI?

- Git init
- Create ZIP on DEV
- Commit
- Create ZIP on QAS
- Commit
- View the diff

SCU0 and /SDF/CD_CCA work quite well for me

**Comment by larshp** [comment]
if you like, maybe a separate repository under https://github.com/abapGit ?

**Comment by abap-weasel** [comment]
Are you familiar with transaction SREPO ("Repository Comparison")? I used it several times for finding differences in landscapes.

**Comment by mbtools** [comment]
I'm aware of the standard tools... but they only go so far (and have the look from the 90s)

I got it working in AG 😄 (took <100 lines of code)

It requires AG developer version and one RFC function on the remote system (+authorization to execute it, of course).

![image](https://user-images.githubusercontent.com/59966492/119227401-eb34b100-bb0d-11eb-90eb-a840a53df82e.png)

In your offline repo, click "import rfc", pick the RFC destination, and the rest works as usual. I haven't tried but it should also work with any table content you include in your repo. And you can pull anything you like from the other system (since we now have selective pull).

**Comment by larshp** [comment]
that's cool 😁 

to me, it seems more like an "import from other system in the landscape" scenario, which is cool, but probably easier using CTS transports. Also the real benefit of using abapGit is having everything versioned in git, which is skipped via RFC.

**Comment by mbtools** [comment]
I was thinking about the other direction: run AG in Dev and import from Q or P (via the RFC). You can revert to the current productive state for example. Since it doesn't depend on CTS transport routes, you could pull from any system reachable via RFC, for example import a package from ERP to BW. 

It allows you to do much more with AG esp if your system is completely offline. 

With great power comes great responsibility. PR on Monday

**Comment by larshp** [comment]
import from Q or P: I'd see that as a security risk, a user running abapGit will need some wide authorizations. Instead I'd suggest scheduling a job to mirror the contents to git(ie. no RFC, so safer).

ERP to BW: yea, but why not just version it in git, using abapGitServer, the user will get more features

**Comment by larshp** [comment]
"Cross system comparison" + "Rollback to previous version" + "Moving/copying objects between systems" are all use-cases which would be nice to have a description of.

They can also be solved in different ways, for abapGit I think the default way to solve these scenarios is using git.

With providing software comes responsibility, I think adding a RFC function module which requires wide authorizations to use can be dangerous if not used properly. And all the scenarios can be solved using git, possibly in a push setup.

I believe that recommending to use git will help users in the long run, having full versioning and a possible setup for performing code reviews.

If the user does not wish to use git, there is classic CTS and SAP standard reports which can already solve some of the scenarios.

**Comment by pokrakam** [comment]
Hmm, while it's a nice idea, I'm also a bit sceptical whether this is good practice. Git is an excellent tool and we're potentially bypassing it completely. There may be some niche scenarios where this may be a minor shortcut, but I agree with Lars that they can all still be solved with Git. 

I see another danger that we're dumbing down VCS too much, in the sense that it makes it too easy for folks to use it as a shortcut to shuffle things around their landscape and never really learn to use Git or even drop it altogether. I see a scenario where one day things go wrong because someone pulls from QA and accidentally wipes a load of work in progress. Git repository? What Git repository? Oops. 
Guess what they'll blame? abapGit. People love blaming tools, my project has completely banned `SE16N` just because someone misused it to edit a table with some bad consequences.

You can set up automated backups to a Git repo. Several systems can even backup to their own branch in the same repo, giving you a wide range of comparison tools. Then in Dev you can pull from your QA branch and still retain complete history and the ability to revert.

**Comment by mbtools** [comment]
I understand the concerns and who can you use tools and how to use them - custom or SAP standard - is always up for debate. For example, you could also create a transport of copies in Prod and import it into Dev (I have seen worse). 

I mentioned before that AG is missing a proper security checks (#3758) but this always ends with "developer has SAP_ALL". So what's the point of talking good/best practices if you don't want to or can't enforce them? (taking away SE16N from a developer? 🙄 ). 

The RFC feature is actually different in that sense since it requires extra authorizations. The security team will have a say whether AG users can "Import via RFC" or not. So there's a bit of good practice. 

The right thing to do is to document the AG usage scenarios and label them with "best practice", "possible alternative" and "exceptions". Let's talk about it on our next call...

PS: If we follow the "oops, no git repo" argument, then we should remove offline repos altogether. 

PPS: I consider it good practice to write-protect all your repos in Dev (and this switch should have an auth check 😉).

**Comment by larshp** [comment]
"developer has SAP_ALL" yea, that's the current state

BUT, we are crossing some lines by providing and encouraging the use of RFC, it then becomes "please create a RFC user with SAP_ALL in production and/or QA, which can be used for abapGit", which can turn into something really bad for the customer

**Comment by pokrakam** [comment]
The way I see it, offline repos fulfil a real need and RFC adds (too much) convenience on top. The effort of Zip files makes Git the preferred alternative. But I feel that a super convenient RFC that is easier to use than a Git repo could be a big temptation given that the majority of ABAP developers are not that familiar with Git.

I agree this would be a great topic to get wider consensus on the next call 👍 

Slightly off the topic, SAP_ALL in DEV has become the exception over my last few projects, I don't think it's that common anymore.

**Comment by flaiker** [comment]
Personally I don't think the authorization part is too bad, in addition to the RFC user with RFC authorization it needs S_DEVELOP etc in read mode but shouldn't need much more. For me it would be much more of a problem arguing that abapGit needs to be in a production system or a test system with real application data (even if the RFC user is in a technical client). abapGit being the largest report ever were some unknown people on the internet call the most private APIs of the SAP system and modify SAP standard db tables directly :P

**Comment by mbtools** [comment]
Agree, the number of times AG will be in productive systems will be limited (any volunteers?). So indeed some of what we discuss is hypothetical. However, there are plenty of non-prod systems or systems of other landscapes that share developments. Maybe you want to compare packages from the SAP namespace to see if you applied the same notes everywhere? 

Sure, SAP systems are their own category. But times are changing. Open source is everywhere (how many SAP systems are running on Linux?). I predict there will be much more of OSS in the ABAP world, too, also in production.

**Comment by flaiker** [comment]
I think it would be a nice feature but the use cases I would mostly use it for (as you mentioned check notes / compare standard packages in an upgrade project, did any transports get stuck or overtaken) are already covered very well by the standard tooling. You would get the comparison for objects that are not versioned in SAP standard but are in abapGit, which would be nice, but you also "lose" the direct integration with the CTS version history, who changed the object when, which transport was it and what is its description, how much was changed, "only whitespace changes" etc. And of course the standard ALV features. So basically I think the feature would be great but for me it already exists with non-git-related tooling so I don't think it would be worth the effort of adding all this to abapGit to make it the better solution without drawbacks. (And of course CCAPPS works on production systems :P )

But I also agree abapGit in production will be something we see more and more. For example if it's a part of the continuous integration pipeline. Or you could even argue the cloud systems have it preinstalled.

**Comment by pokrakam** [comment]
> I don't think users will all of a sudden use offline repos with RFC and skip Git. They are smart enough to understand the use cases and benefits of using Git repos.

I think this is where we have different experiences, because I'm coming from the exact opposite angle. For many people abapGit is nothing more than a tool to upload and download code and fetch stuff from the internet. The majority of developers I speak to and work with have barely heard of Git. Not to say they aren't smart, but even once you start working with Git it can take some time to get your head around properly (I still haven't). 

Recently I tried to introduce abapGit to some smart systems managers and basis folk as a way to manage parallel development systems, and among the first questions was why we need a Git server, can't we go direct between systems? 
And not that long ago I encountered a setup where developers pushed their work to QA via TOCs and STMS, all in the name of avoiding transport conflicts. I hope that explains my more cautious approach... we're already handing people a loaded gun, let's not remove the safety catch :-)

Maybe a compromise is to make RFC readonly?

> Happy to discuss more with a broader audience.

Agree. 🙂
